kubernetes-sigs/aws-alb-ingress-controller
# finalize user-facing elements(finalizer/tagging/labeling)

# finalize all user-facing elements, details below:
## AWS
### For AWS resources created by this controller, the tagging strategy is as follows:
 * `elbv2.k8s.aws/cluster: cluster-name` will be applied on all AWS resources.
 * `ingress.k8s.aws/stack: stack-id` will be applied on all AWS resources provisioned for Ingress resources:
   * For explicit IngressGroup, `stack-id` will be `groupName`
   * For implicit IngressGroup, `stack-id` will be `namespace/ingressName`
 * `ingress.k8s.aws/resource: resource-id` will be applied on all AWS resources provisioned for Ingress resources:
   * For LoadBalancer, `resource-id` will be `LoadBalancer`
   * For Managed LB SecurityGroup, `resource-id` will be `ManagedLBSecurityGroup`
   * For TargetGroup, `resource-id` will be `namespace/ingressName-serviceName:servicePort`
 * `service.k8s.aws/stack: stack-id` will be applied on all AWS resources provisioned for Service resources:
   * `stack-id` will be `namespace/serviceName`
 * `service.k8s.aws/resource: resource-id` will be applied on all AWS resources provisioned for Service resources:
   * For LoadBalancer, `resource-id` will be `LoadBalancer`
   * For TargetGroup, `resource-id` will be `namespace/serviceName:servicePort`
### For AWS resources created by this controller, the naming strategy is as follows:
1. LoadBalancer:
   * For explicit IngressGroup: `k8s-%.17s{groupName}-%.10s{hash}`
   * For implicit IngressGroup & Service: `k8s-%.8s{namespace}%.8s{name}-%.10s{hash}`
2. TargetGroup: `k8s-%.8s{namespace}%.8s{name}-%.10s{hash}`
## K8s
### For K8s resources created by this controller, the labelling strategy is as follows:
Note: We break "stack" into "stack-namespace" and "stack-name" for  implicit IngressGroup and Service since label value must be within 63 characters.
 * For explicit IngressGroup, the following tags will be applied on all K8s resources:
   * `ingress.k8s.aws/stack: groupName`
 * For implicit IngressGroup, the following tags will be applied on all K8s resources: 
   * `ingress.k8s.aws/stack-namespace: namespace`
   * `ingress.k8s.aws/stack-name: ingressName`
 * For Service, the following tags will be applied on all K8s resources:
   * `service.k8s.aws/stack-namespace: namespace`
   * `service.k8s.aws/stack-name: serviceName`
### For K8s resources, the following finalizers are used:
* For Ingresses within explicit IngressGroup: `group.ingress.k8s.aws/groupName`
* For Ingresses don't belong to explicit IngressGroup: `ingress.k8s.aws/resources`
* For Services: `service.k8s.aws/resources`
* For TargetGroupBindings: `elbv2.k8s.aws/resources`
### For pods behind targetGroupBindings with IP targetType, the following readinessGate will be injected:
* `target-health.elbv2.k8s.aws/targetGroupBindingName`

# additional changes:
1. support Tag on creation for TargetGroups.
2. preserve old tags on AWS resources provisioned by AWSALBLoadBalancerController(v1.1.3+), so that downgrade is support. 
3. normalize the namespace/name used in resource names. so that targetGroup won't be named like "k8s-xxxx--dde"(two continuous '-' within name).
4. standard all naming generation for resources created for Ingress and Service, especially:
    1. the LoadBalancer name of both ALB/NLB follows the same `k8s-.8{namespace}-.8{name}-.10{hash}` pattern. For IngressGroups, the name can be `k8s-.17{groupName}-.10{hash}`
    2. the TargetGroups are now named following similar pattern across ALB/NLB.
5. the resourceID of LoadBalancer for NLB stack is changed to `LoadBalancer`
6. use shared k8s.finalizerManager to handle finalizer logic for IngressGroup
7. abstract a TaggingManager for EC2 Tagging operations.
8. various code style corrections

# Test done:
1. create Ingress -> delete Ingress   (success)
2. create IngressGroup -> delete IngressGroup  (success)
3. create Ingress with long name  in long namespace -> delete Ingress  (success)
4. create Ingress with old controller -> disable old controller -> run new controller -> verify resource state 
    -> disable new controller -> delete TargetGroupBindings -> run old controller -> verify resource state  (success)
   
Fixes #1463
